### PR TITLE
add dims keyword for `padded_views` and `sym_paddedviews`

### LINKED ISSUE #48: Dims Keyword addition
Hey @johnnychen94 ,I think this works but it looks clumsy.I would need some guidance:)question is where we apply dimension keyword effect which tells what to use and what not to.
Pluto Notebook: https://github.com/ashwani-rathee/Demos/blob/main/PaddedViewsIssue.jl
Pdf Version: https://github.com/ashwani-rathee/Demos/blob/main/%F0%9F%8E%88%20PaddedViewsIssue.pdf
```
function PaddedView(fillvalue::FT, data::AbstractArray{T}, args...) where {FT, T}
    PaddedView{filltype(FT, T)}(fillvalue, data, args...)
end

function PaddedView{FT}(fillvalue,data::AbstractArray{T,N},
			sz::Tuple{Integer,Vararg{Integer}},dims=1) where {FT,T,N}
	if(dims==0)
	  val=copy(sz[1])
	 new_tuple=(size(data)[2],val)
	elseif(dims==1)
	  val=copy(sz[2])
	 new_tuple=(val,size(data)[1])
	end			
        inds = map(OneTo, new_tuple)     
    PaddedView{FT,N,typeof(inds),typeof(data)}(convert(FT, fillvalue), data, inds)
end
```
Before applying dims like this in function,I applied it to the struct PaddedView too,but I am not sure which should we use..
I use it on this:
```
img_source = testimage("woman_darkhair")
pyramid = gaussian_pyramid(img_source, 5, 2, 1)
PaddedView(-1, pyramid[2], (512,512),0) # 0 if we want update in rows direction, and 1 if we want to update in columns direction
```

### PR DESCRIPTION
**Pull request by ashwanirathee**
@johnnychen94 Initial PR on addition of keyword dims in padded views and also in sym_padded_views later

closes #48

### CODE REVIEW

File: src/PaddedViews.jl [hunk 1]
```diff
@@ -219,7 +219,23 @@ julia> a1p[CartesianIndices(a1)]
 """
 function paddedviews(fillvalue, As::AbstractArray...)
     inds = outerinds(As...)
-    map(A->PaddedView(fillvalue, A, inds), As)
+    dims1 = 1
+    dims2 = (1,2)
+    dims3 = 1:2
```
Comment (johnnychen94) [inline src/PaddedViews.jl]: To be clear, `PaddedView` handles arrays of arbitrary dimensions, for example, `As` can be a list of arrays with dimension N=4. I don't think the current code handles `N>2` case.

File: src/PaddedViews.jl [hunk 2]
```diff
@@ -219,7 +219,23 @@ julia> a1p[CartesianIndices(a1)]
 """
 function paddedviews(fillvalue, As::AbstractArray...)
```
Comment (johnnychen94) [inline src/PaddedViews.jl]: we need to add a keyword `dims` here

**Review by johnnychen94 — COMMENTED** [review]
Generally looks good, except that the current implementation is 2D specific; we actually want to support arbitrary dimensions.

I suggest that you write down some tests in `test/runtest.jl` and test it via

```bash
julia --project=. -e "using Pkg; Pkg.test()"
```

This helps you become confident in your implementation.

Also, when the functionality is done, don't forget to add documentation, and update `README.md`.

File: src/PaddedViews.jl (hunk 1 above)
Comment (ashwanirathee) [reply src/PaddedViews.jl]: @johnnychen94 Can the person provide the tuple in like in any random fashion..like can I provide it in following ways:
- (1,2,4,5,3) - I need sort and unique for this
- (1,2,3,4,5)- if it's continous and properly sorted,then it won't be a issue
- (1,1,1) - also this case
- (0,1,2,...) - also this case

Comment (johnnychen94) [reply src/PaddedViews.jl]: I'll follow the cat convention:

```julia
julia> cat([1,2], [3 4]; dims=(2,1))
3×3 Array{Int64,2}:
 1  0  0
 2  0  0
 0  3  4
```

Comment (ashwanirathee) [reply src/PaddedViews.jl]: Also we get dimension mismatch issue when arrays data is returned,which is mosaicviews data handling issue but that does arise for some reason I don't get yet...though we won't be talking about it here,right?

Comment (johnnychen94) [reply src/PaddedViews.jl]: The dimension mismatch issue you mentioned in #48 is not very clear to me, they should be unrelated. My best guess is that once we consolidate this feature in the PaddedViews side, the issue got fixed automatically.

FYI, there's a PR (https://github.com/JuliaArrays/MosaicViews.jl/pull/31) related to the `_padded_cat` you're referring to, so no, I'm not going to check it until we get the current PaddedViews PR done.

**Comment by ashwanirathee** [comment]
Maybe the JuliaFormatter messed it all up a little,I formatted the whole code and now we have conflicts 😓 .I formatted it and it itself formatted too the code in vs code whenever save button is pushed.

**Comment by johnnychen94** [comment]
Yeah... could you revert the unrelated changes? It's a little bit mess to review the diffs now...

**Comment by ashwanirathee** [comment]
I think this does the job for the paddedviews, next is sym_ paddedviews

**Comment by johnnychen94** [comment]
Didn't check very carefully, but it looks good to me so far. Will recheck it when `sym_paddedviews` is supported.

**Comment by ashwanirathee** [comment]
without collect, it throws errors, Needs to check why

File: test/runtests.jl [hunk 3]
```diff
@@ -156,6 +156,96 @@ end
...
+    a1 = reshape([1, 2, 3,4,5,6,7,8,9], 3, 3)
+    a2 = reshape([5, 6], 2, 1)
+    a2b = [-1 5 -1; -1 6 -1; -1 -1 -1]
+    a2c = reshape([5, 6, -1], 3, 1)
+    a2r =  [-1 5 -1; -1 6 -1]
+
+    a1f, a2f = sym_paddedviews(-1, a1, a2;dims=(1))
+    @test a2f == a2c
+    a1f, a2f = sym_paddedviews(-1, a1, a2;dims=(2))
+    @test collect(a2f) == a2r
```
Comment (johnnychen94) [inline test/runtests.jl]: > without collect, it throws errors, Needs to check why

`PaddedArray` might not be 1-based Array, to properly compare two arrays, you need to make sure their axes are the same.

```julia
# from `@less ==(rand(4, 4), rand(4, 4))
function (==)(A::AbstractArray, B::AbstractArray)
    if axes(A) != axes(B)
        return false
    end
    anymissing = false
    for (a, b) in zip(A, B)
        eq = (a == b)
        if ismissing(eq)
            anymissing = true
        elseif !eq
            return false
        end
    end
    return anymissing ? missing : true
end
```

**Review by johnnychen94 — APPROVED** [review]
LGTM, can you also update the docstrings and README examples? When that's done I think it's okay to merge this.

**Pull request merged**